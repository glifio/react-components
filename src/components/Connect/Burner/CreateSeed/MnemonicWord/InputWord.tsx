import styled, { css } from 'styled-components'
import PropTypes from 'prop-types'
import { useEffect, useState } from 'react'
import { Colors } from '../../../../theme'
import { MnemonicWord } from './MnemonicWord'

enum InputWordState {
  Input = 0,
  Correct,
  Incorrect
}

// Styles need the !important declaration to
// override the base-css text input styles
const InputWordEl = styled.input`
  display: block !important;
  width: 100% !important;
  padding: 0 1em !important;
  border: 2px solid !important;
  border-radius: 1em !important;
  text-align: center !important;

  ${props => {
    switch (props.state) {
      case InputWordState.Correct:
        return css`
          color: ${Colors.WHITE} !important;
          border-color: ${Colors.GREEN_MEDIUM} !important;
          background-color: ${Colors.GREEN_MEDIUM} !important;
        `
      case InputWordState.Incorrect:
        return css`
          color: ${Colors.WHITE} !important;
          border-color: ${Colors.RED_MEDIUM} !important;
          background-color: ${Colors.RED_MEDIUM} !important;
        `
      default:
        return css`
          color: ${Colors.PURPLE_MEDIUM} !important;
          border-color: ${Colors.PURPLE_MEDIUM} !important;
          background-color: ${Colors.PURPLE_LIGHT} !important;
        `
    }
  }}

  &:focus {
    color: ${Colors.PURPLE_MEDIUM} !important;
    border-color: ${Colors.PURPLE_MEDIUM} !important;
    background-color: ${Colors.WHITE} !important;
  }
`

export const InputWord = ({ num, word, onValidChange }: InputWordProps) => {
  const [inputWord, setInputWord] = useState<string>('')
  const [state, setState] = useState<InputWordState>(InputWordState.Input)
  const [valid, setValid] = useState<boolean>(false)

  useEffect(() => setValid(word === inputWord), [word, inputWord])
  useEffect(() => onValidChange(valid), [onValidChange, valid])

  return (
    <MnemonicWord num={num}>
      <InputWordEl
        type='text'
        size={1}
        state={state}
        value={inputWord}
        onChange={e => setInputWord(e.target.value)}
        onFocus={() => setState(InputWordState.Input)}
        onBlur={() =>
          setState(valid ? InputWordState.Correct : InputWordState.Incorrect)
        }
      />
    </MnemonicWord>
  )
}

interface InputWordProps {
  num: number
  word: string
  onValidChange: (valid: boolean) => void
}

InputWord.propTypes = {
  num: PropTypes.number.isRequired,
  word: PropTypes.string.isRequired,
  onValidChange: PropTypes.func.isRequired
}
