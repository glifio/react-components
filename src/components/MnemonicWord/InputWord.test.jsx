import { render } from '@testing-library/react'

import { Base } from './InputWord.stories'

describe('Box', () => {
  test('renders the story', () => {
    const mockSetCorrectWordCount = jest.fn()
    const { container } = render(
      <Base {...Base.args} setCorrectWordCount={mockSetCorrectWordCount} />
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
