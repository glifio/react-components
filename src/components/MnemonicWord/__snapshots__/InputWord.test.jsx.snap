// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box renders the story 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  -webkit-flex: 1 0 10em;
  -ms-flex: 1 0 10em;
  flex: 1 0 10em;
}

.c1 {
  width: 1.5em;
  height: 1.5em;
  font-size: 1.125em;
  text-align: center;
  line-height: 1.5em;
}

.c2 {
  background-color: transparent;
  color: #5E26FF;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 32px;
  border-radius: 100px;
  border: 1px solid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
  font-size: 1.25rem;
}

.c2:focus {
  box-shadow: 0;
  outline: 0;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    12
  </div>
  <input
    class="c2"
    color="core.primary"
    display="flex"
    font-size="3"
    height="5"
    value=""
    width="100%"
  />
</div>
`;
